### ISSUE TITLE
Notebook: make attention a neutral switchable option (SDPA default / opt-in FA2) + even out section-doc density (Metrics etc.)

### ISSUE THREAD

**Issue by hiouchiy** [hiouchiy/air-sample-bert#6]
Two related notebook-doc improvements, per maintainer review.

---

## Part A — Make the attention implementation a neutral, switchable **option** (01 & 02)

**Current state (`01_finetune_singlegpu.py` L177–183, same pattern in `02_finetune_multigpu.py` L161–167):** `_pick_attn_implementation()` auto-uses `flash_attention_2` **if `flash_attn` imports**, else `sdpa`. There is **no parameter to control it**, and the §5 markdown (L161–163) reads as a defensive justification:

> "…Flash Attention 2 … has known numerical-stability edge cases with some `flash-attn` builds, so we **fall back** to PyTorch SDPA … This keeps the demo **robust**."

This isn't neutral, it's not actually SDPA-by-default (it prefers FA2 whenever installed), and a reader can't choose.

**Proposed:** present both as options, default to **SDPA**, let users opt into FA2 via an env var, and show how.

*§5 markdown (neutral):*
> This model supports two attention backends:
> - **PyTorch SDPA** (`sdpa`) — built into PyTorch, no extra install, runs anywhere. **Default in this sample.**
> - **Flash Attention 2** (`flash_attention_2`) — faster / more memory-efficient on A10/H100, but it ships as a CUDA extension that must be **compiled to match your exact PyTorch + CUDA + GPU architecture**; a matching prebuilt wheel often isn't available, so install can take many minutes or fail.
>
> To use FA2: `%pip install flash-attn --no-build-isolation` and set `ATTN_IMPLEMENTATION=flash_attention_2`.

*Code:*
```python
# Config
attn_implementation: str = _env("ATTN_IMPLEMENTATION", "sdpa")  # "sdpa" (default) | "flash_attention_2"
# build_model
model = AutoModelForSequenceClassification.from_pretrained(..., attn_implementation=cfg.attn_implementation)
```
This removes the "fallback"/"keeps the demo robust"/"numerical-stability" wording, makes the default deterministic, and lets the user switch. Apply the same to `02_finetune_multigpu.py`.

*(Note: "fallback" is neutral in native English technical writing, but framing it as a user-selectable option avoids the word entirely and reads better.)*

---

## Part B — Even out section-level doc density (the "Metrics" observation)

The notebooks are **uneven**: some cells over-explain with self-justifying prose, while others are a bare heading with **no description at all**. Pick one concise functional line per section and drop the editorializing.

**Bare headings (no body) — add a one-line "what this cell does":**
- `01_finetune_singlegpu.py` §4 **Metrics** (L142) — e.g. "Compute accuracy and **macro-F1** (weights all four classes equally)." (also explains *why* macro-F1)
- `03_batch_inference.py` §1 (L21), §4 Load input rows (L110), §5 (L126), §7 (L176)

**Self-justifying prose to trim (customer doesn't need the rationale):**
- `01_finetune_singlegpu.py` L105–106 — "…no Spark or Unity Catalog table is required, which keeps the example fully reproducible on any AI Runtime GPU." → keep just "AG News is a public dataset that ships with the `datasets` library."
- `01_finetune_singlegpu.py` L163 — "This keeps the demo robust." (removed by Part A anyway)

**Keep (genuinely educational):** the multi-GPU "why this pattern" paragraph in `02`; the brief "plain file write — no Spark" note in `03` (pre-empts the Spark log noise).

Net: uniform, useful, non-editorializing section docs — fixing both the "too chatty" cells and the empty ones in one pass.

**Comment by hiouchiy** [comment]
Reopening for **Part C only** — Parts A (attention as a neutral, switchable option) and B (section-doc density) are done and verified in `docs(notebook): make notebooks notebook-native + attach steps + doc cleanup` (3483157). Thanks!

**Still open — Part C: explain the `.to("cpu")` before logging.** No comment/markdown was added, so in this GPU sample the model is still moved to CPU with no explanation:
- `01_notebook/01_finetune_singlegpu.py` L273 — `model=trainer.model.to("cpu")`
- `01_notebook/02_finetune_multigpu.py` L234 — same

A reader hits `.to("cpu")` in a GPU notebook and reasonably wonders why. Suggested one-line comment (both files), matching how the XGBoost sample documents its equivalent:
```python
# Package on CPU so the logged model loads on any hardware, and so MLflow's signature
# inference (it runs the pipeline once at log time) doesn't hit a GPU/CPU device mismatch.
# Training ran on GPU; only this serialization step uses CPU.
clf = pipeline("text-classification", model=trainer.model.to("cpu"), tokenizer=tokenizer)
```
(Optionally add one sentence to the §6 "Log to MLflow & register" markdown.) Full rationale in the earlier Part C comment above.

**Comment by hiouchiy** [comment]
Fixed in bb64bbd. Part C done: added a comment (01 & 02, notebook + CLI) explaining why the pipeline is packaged on CPU before logging — portable artifact + avoids a GPU/CPU device mismatch during MLflow's signature inference; training still runs on GPU. (Parts A & B were already done in 3483157.)

**Comment by hiouchiy** [comment]
## Notebook/CLI drift: the attention change was applied to the notebook but NOT the CLI

A notebook-vs-CLI diff shows the switchable-attention design from this issue was applied to the **notebook** only:
- `01_notebook/01_finetune_singlegpu.py` & `02_finetune_multigpu.py`: use `attn_implementation = _env("ATTN_IMPLEMENTATION", "sdpa")` (the neutral, SDPA-default, env-switchable knob). 
- `02_cli/01_finetune_singlegpu.py` & `02_finetune_multigpu.py`: **still use the old `_pick_attn_implementation()` auto-detect** (`try import flash_attn → flash_attention_2 except → sdpa`).

So the two forms now behave differently (notebook = deterministic SDPA default + opt-in; CLI = auto-prefer FA2 if importable). Apply the same `ATTN_IMPLEMENTATION` knob to the `02_cli/` copies so notebook and CLI match. (Found while checking notebook↔CLI consistency; the rest of the scripts differ only in expected scaffolding.)
